### PR TITLE
Graphql websocket

### PR DESCRIPTION
**Pull request by sharadregoti**

### CODE REVIEW

**Comment by YourTechBud** [comment]
@sharadregoti you need to resolve this merge conflicts. Let me know once you are done

File: utils/server/websocket.go [hunk 1]
```diff
@@ -152,3 +162,144 @@ func (s *Server) handleWebsocket() http.HandlerFunc {
...
+func (s *Server) handleGraphqlSocket(adminMan *admin.Manager) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		project := mux.Vars(r)["project"]
+		socket, err := upgrader.Upgrade(w, r, nil)
+		if err != nil {
+			log.Println("upgrade:", err)
+			return
+		}
+		defer socket.Close()
+
```
Comment (YourTechBud) [inline utils/server/websocket.go]: Make a unique string named `clientID` here. Use UUID.

File: utils/server/websocket.go [hunk 2]
```diff
@@ -152,3 +162,144 @@ func (s *Server) handleWebsocket() http.HandlerFunc {
...
+				data.Token = token
+				data.DBType = v.Directives[0].Name.Value
+				data.Project = project
+				data.Group = v.Name.Value
+				data.Type = m.Type
+				data.ID = m.ID
+				graphqlIDMapper.Store(m.ID, data.Group)
+
+				// Subscribe to realtime feed
+				feedData, err := s.realtime.Subscribe(ctx, m.ID, s.auth, s.crud, data, func(feed *model.FeedData) {
```
Comment (YourTechBud) [inline utils/server/websocket.go]:
```suggestion
				feedData, err := s.realtime.Subscribe(ctx, clientID, s.auth, s.crud, data, func(feed *model.FeedData) {
```

File: utils/server/websocket.go [hunk 3]
```diff
@@ -152,3 +162,144 @@ func (s *Server) handleWebsocket() http.HandlerFunc {
...
+				// For realtime subscribe event
+				group, ok := graphqlIDMapper.Load(m.ID)
+				if !ok {
+					return
+				}
+				data := new(model.RealtimeRequest)
+				data.ID = m.ID
+				data.Group = group.(string)
+
+				s.realtime.Unsubscribe(m.ID, data)
```
Comment (YourTechBud) [inline utils/server/websocket.go]:
```suggestion
				s.realtime.Unsubscribe(clientID, data)
```

File: utils/server/websocket.go (hunk 1 above)
Comment (YourTechBud) [reply utils/server/websocket.go]: also call `defer s.realtime.RemoveClient(clientID)`

File: utils/constants.go [hunk 4]
```diff
@@ -245,6 +245,29 @@ const (
 	Subscribe PubsubType = "subscribe"
 )
 
+const (
+	// GqlConnectionKeepAlive send every 20 second to client over websocket
+	GqlConnectionKeepAlive = "ka" // Server -> Client
```
Comment (YourTechBud) [inline utils/constants.go]: All these are not matching the types specified in the protocol. Everything should be in capital.

File: utils/server/websocket.go [hunk 5]
```diff
@@ -152,3 +162,144 @@ func (s *Server) handleWebsocket() http.HandlerFunc {
...
+		ticker := time.NewTicker(20 * time.Second)
+		defer ticker.Stop()
+
+		for {
+
+			m := graphqlMessage{}
+			if err := socket.ReadJSON(m); err != nil {
+				return
+			}
+			select {
```
[the remark this answers is not in the document]
Comment (sharadregoti) [reply utils/server/websocket.go]: ok, but how do i pass message id,(m.id)
i can pass the status only GqlConnectionKeepAlive

File: utils/server/websocket.go [hunk 6]
```diff
@@ -152,3 +162,144 @@ func (s *Server) handleWebsocket() http.HandlerFunc {
...
+				source := source.NewSource(&source.Source{
+					Body: []byte(m.Payload.Query),
+				})
+				// parse the source
+				doc, err := parser.Parse(parser.ParseParams{Source: source})
+				if err != nil {
+					channel <- (&graphqlMessage{ID: m.ID, Payload: payloadObject{Status: utils.GqlConnectionError, Error: err.Error()}})
+					return
+				}
+				v := doc.Definitions[0].(*ast.OperationDefinition).SelectionSet.Selections[0].(*ast.Field)
```
[the remark this answers is not in the document]
Comment (sharadregoti) [reply utils/server/websocket.go]: for every type assertions or the last one?

[the remark this answers is not in the document]
Comment (YourTechBud) [reply utils/server/websocket.go]: For everyone

File: utils/server/websocket.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (YourTechBud) [reply utils/server/websocket.go]: In that case you'll have to create a ticker for each start. And make sure you close them all when websocket disconnects

[the remark this answers is not in the document]
Comment (YourTechBud) [reply utils/server/websocket.go]: Or maybe, you already have a map for all the query ids. Loop over it and send a message for each one

[the remark this answers is not in the document]
Comment (sharadregoti) [reply utils/server/websocket.go]: or can i make a variable which make sure that the go func gets executed only once
isOnlyOnce = true 
if isOnlyOnce {
isOnlyOnce = false
go func 
}

[the remark this answers is not in the document]
Comment (YourTechBud) [reply utils/server/websocket.go]: You'll need to figure out a way for this.

[the remark this answers is not in the document]
Comment (sharadregoti) [reply utils/server/websocket.go]: range over ticker.C is not possible, only select can be done.
if we use range then go routine cannot be closed even if the ticker is stopped

File: utils/server/websocket.go [hunk 7]
```diff
@@ -247,58 +267,75 @@ func (s *Server) handleGraphqlSocket(adminMan *admin.Manager) http.HandlerFunc {
...
 				data := new(model.RealtimeRequest)
 				data.ID = m.ID
 				data.Group = group.(string)
 
-				s.realtime.Unsubscribe(m.ID, data)
-				channel <- (&graphqlMessage{ID: m.ID, Payload: payloadObject{Status: utils.GqlStop}})
-
+				s.realtime.Unsubscribe(clientID, data)
+				channel <- (&graphqlMessage{ID: m.ID, Payload: payloadObject{Status: utils.GQL_STOP}})
+				graphqlIDMapper.Delete(m.ID)
```
Comment (sharadregoti) [inline utils/server/websocket.go]: should I delete the id after unsubscribe

File: utils/server/websocket.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (YourTechBud) [reply utils/server/websocket.go]: Stopping the ticker automatically closes the channel right?

File: utils/server/websocket.go (hunk 7 above)
Comment (YourTechBud) [reply utils/server/websocket.go]: Yes

File: utils/server/websocket.go [hunk 8]
```diff
@@ -152,3 +164,189 @@ func (s *Server) handleWebsocket() http.HandlerFunc {
...
+
+		closeConnAliveRoutine := make(chan bool)
+		defer close(closeConnAliveRoutine)
+
+		go func() {
+			for res := range channel {
+				err := socket.WriteJSON(res)
+				if err != nil {
+					log.Println(err)
+					return
```
Comment (YourTechBud) [inline utils/server/websocket.go]: Don't return here

File: utils/server/websocket.go [hunk 9]
```diff
@@ -152,3 +164,189 @@ func (s *Server) handleWebsocket() http.HandlerFunc {
...
+				}
+			}
+		}()
+
+		for {
+
+			m := graphqlMessage{}
+			if err := socket.ReadJSON(m); err != nil {
```
Comment (YourTechBud) [inline utils/server/websocket.go]:
```suggestion
			if err := socket.ReadJSON(&m); err != nil {
```

File: utils/server/websocket.go [hunk 10]
```diff
@@ -152,3 +164,189 @@ func (s *Server) handleWebsocket() http.HandlerFunc {
...
+								return
+							case <-ticker.C:
+								channel <- &graphqlMessage{ID: m.ID, Payload: payloadObject{Status: utils.GQL_CONNECTION_KEEP_ALIVE}}
+							}
+						}
+					}()
+				}
+
+			case utils.GQL_START:
+				data := new(model.RealtimeRequest)
```
Comment (YourTechBud) [inline utils/server/websocket.go]: Can you create data below and assign all its fields while creating it?

File: utils/server/websocket.go [hunk 11]
```diff
@@ -152,3 +164,189 @@ func (s *Server) handleWebsocket() http.HandlerFunc {
...
+					channel <- &graphqlMessage{ID: feed.QueryID, Payload: payloadObject{Status: utils.GQL_DATA, Data: graphqlSuccessData{Type: feed.Type, Doc: feed.Payload, DocID: feed.DocID}}}
+				})
+
+				if err != nil {
+					channel <- &graphqlMessage{ID: m.ID, Payload: payloadObject{Status: utils.GQL_ERROR, Error: []gqlError{{Message: err.Error()}}}}
+					closeConnAliveRoutine <- true
+					return
+				}
+
+				channel <- (&graphqlMessage{ID: m.ID, Payload: payloadObject{Status: utils.GQL_START, Data: feedData}})
```
Comment (YourTechBud) [inline utils/server/websocket.go]: you can't just pass feedData like this. You'll have to iterate over it and make a `graphqlSuccessData` for each one of them and pass that in the channel in the loop itself

**Pull request merged**